### ISSUE TITLE
[Feature Request] Flattened indices option for max pooling

### ISSUE THREAD

**Issue by ananiask8** [pytorch/pytorch#18796]
## 🚀 Feature
Option to return flattened indices for max pooling layers.

## Motivation
Some operations require to have access to the flattened indices that ONNX returns initially. For instance, let's consider this vision example: I am working on a layer that counts how many times the element (pixel) at position `i` was the winner (selected as maximum) in a max pooling 2d operation.

As things are right now, this is the solution:
```python
def forward(self, input):
    _, indices = self.maxpool2d(input)
    N, C, H, W = input.size()
    minlength = H * W
    Nout, Cout, Hout, Wout = indices.size()
    indices_count = [t.bincount(minlength=minlength) for t in indices.view(-1, Hout * Wout).unbind()]
    return torch.stack(indices_count).view(input.size())
```

This implementation is very slow. The overhead introduced by running this for-loop could be avoided if flattened indices were returned. This is how the code would look like then:
```python
def forward(self, input):
    input_flattened_length, input_shape = input.nelement(), input.size()
    _, indices = self.maxpool2d(input)
    return indices.view(-1).bincount(minlength=input_flattened_length).view(input_shape)
```

The indices for max pooling 2d are currently referencing local frames, non-flattened. See [this issue](https://github.com/pytorch/pytorch/pull/16455#issuecomment-460776407) for a clearer picture of what this means.

## Pitch
This feature would allow to return flattened indices, in the same way as [tf.nn.max_pool_with_argmax](https://www.tensorflow.org/api_docs/python/tf/nn/max_pool_with_argmax) does.

## Alternatives
1. To implement `apply_along_axis`. This is a draft showing how it would solve my problem:
```python
def forward(self, input):
    _, indices = self.maccpool2d(-input)
    N, C, H, W = input.size()
    minlength = H * W
    Nout, Cout, Hout, Wout = indices.size()
    r = np.apply_along_axis(lambda x: np.bincount(x, minlength=minlength), axis=-1, arr=indices.view(N, C, -1).numpy())
    return torch.from_numpy(r).view(input.size())
```
2. To add a frequency counter that takes the indices returned by max pooling 2d (for instance) and sets in the location of the tensor corresponding to each index, the amount of times it was encountered in the returned indices.

## Additional context
In short, what I am trying to implement is the following:

![CodeCogsEqn (2)](https://user-images.githubusercontent.com/7946422/55476426-d1c30900-5616-11e9-85ba-4ae3f2a91677.gif)
where C contains the unfolded set of kernel windows.

In words: for each kernel window `c`, it adds `1` in the output at the location of the winner, `1[argmax c]`.

**Comment by dashesy** [comment]
Can you add (using broadcast) the vector `[[0], [w], [2 * w], ...]` (with `w` being the width) to the indices tensor, then it will be flattened indices.

**Comment by ananiask8** [comment]
Thank you very much @dashesy for the suggestion. I got it working like that for the time being. Before closing this thread I would like to ask if this could be done more efficiently by other methods, such as `unfold` and `fold`.
